File: routes/utils/premium_content.py
"""
Premium Content Utility Functions

This module provides utilities for handling premium content filtering,
content truncation, and premium status checking.
"""

import re
from typing import Tuple, Optional
from flask import current_app
from flask_login import current_user


def is_premium_user() -> bool:
    """
    Check if the current user has active premium access.
    
    Returns:
        bool: True if user has active premium subscription, False otherwise
    """
    if not current_user.is_authenticated:
        return False
    
    return current_user.has_active_premium_subscription()


def should_show_premium_content(content_is_premium: bool) -> bool:
    """
    Determine if premium content should be shown to the current user.
    
    Args:
        content_is_premium (bool): Whether the content is marked as premium
        
    Returns:
        bool: True if content should be shown, False if it should be truncated
    """
    if not content_is_premium:
        return True  # Non-premium content is always shown
    
    return is_premium_user()


def truncate_markdown_content(content: str, max_words: int = 150) -> Tuple[str, bool]:
    """
    Truncate markdown content to a specified number of words while preserving markdown structure.
    
    Args:
        content (str): The markdown content to truncate
        max_words (int): Maximum number of words to show
        
    Returns:
        Tuple[str, bool]: (truncated_content, was_truncated)
    """
    if not content:
        return "", False
    
    # Split content into words while preserving markdown structure
    words = re.findall(r'\S+|\n+', content)
    
    if len(words) <= max_words:
        return content, False
    
    # Take first max_words words
    truncated_words = words[:max_words]
    
    # Reconstruct content
    truncated_content = ''.join(truncated_words)
    
    # Add ellipsis if content was truncated
    if truncated_content and not truncated_content.endswith('\n'):
        truncated_content += '...'
    
    return truncated_content, True


def process_premium_content(content: str, is_premium_content: bool, max_words: int = 150) -> Tuple[str, bool, bool]:
    """
    Process content based on premium status and user access.
    
    Args:
        content (str): The original content
        is_premium_content (bool): Whether the content is premium
        max_words (int): Maximum words to show for non-premium users
        
    Returns:
        Tuple[str, bool, bool]: (processed_content, is_truncated, show_premium_notice)
    """
    if not should_show_premium_content(is_premium_content):
        # User doesn't have premium access for premium content
        truncated_content, was_truncated = truncate_markdown_content(content, max_words)
        return truncated_content, was_truncated, True
    
    # User has access to full content
    return content, False, False


def get_premium_content_stats(content: str, is_premium_content: bool) -> dict:
    """
    Get statistics about premium content processing.
    
    Args:
        content (str): The original content
        is_premium_content (bool): Whether the content is premium
        
    Returns:
        dict: Statistics about the content
    """
    if not content:
        return {
            'total_words': 0,
            'total_chars': 0,
            'is_premium': is_premium_content,
            'user_has_access': should_show_premium_content(is_premium_content)
        }
    
    words = re.findall(r'\S+', content)
    chars = len(content)
    
    return {
        'total_words': len(words),
        'total_chars': chars,
        'is_premium': is_premium_content,
        'user_has_access': should_show_premium_content(is_premium_content)
    }


def generate_premium_notice_html(content_type: str = "artikel") -> str:
    """
    Generate HTML for premium content notice.
    
    Args:
        content_type (str): Type of content (artikel, bab, etc.)
        
    Returns:
        str: HTML for premium notice
    """
    return f"""
    <div class="premium-content-notice bg-gradient-to-r from-primary/10 to-secondary/10 border border-primary/20 rounded-lg p-6 my-6">
        <div class="flex items-center gap-3 mb-4">
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-6 h-6 text-primary">
                <path fill-rule="evenodd" d="M5.166 2.621v.858c-1.035.148-2.059.33-3.071.543a.75.75 0 00-.584.859 6.937 6.937 0 006.222 6.222.75.75 0 00.859-.584 48.332 48.332 0 001.243-5.034 48.332 48.332 0 005.034-1.243.75.75 0 00.584-.859 6.937 6.937 0 00-6.222-6.222.75.75 0 00-.859.584 48.332 48.332 0 00-1.243 5.034zM13.5 2.25a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V3a.75.75 0 01.75-.75zM13.5 6.75a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V7.5a.75.75 0 01.75-.75zM13.5 11.25a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V12a.75.75 0 01.75-.75z" clip-rule="evenodd" />
            </svg>
            <h3 class="text-lg font-semibold text-foreground">Konten Premium</h3>
        </div>
        <p class="text-muted-foreground mb-4">Untuk membaca {content_type} ini secara lengkap, Anda memerlukan langganan premium.</p>
        <div class="flex flex-wrap gap-3">
            <a href="/premium" class="button button-primary">
                Berlangganan Premium
            </a>
            <a href="/login" class="button button-outline">
                Login
            </a>
        </div>
    </div>
    """


def create_content_mask_html(content_id: str, content_type: str = "artikel") -> str:
    """
    Create HTML for content mask overlay.
    
    Args:
        content_id (str): Unique identifier for the content
        content_type (str): Type of content
        
    Returns:
        str: HTML for content mask
    """
    return f"""
    <div id="content-mask-{content_id}" class="content-mask-overlay">
        <div class="content-mask-backdrop"></div>
        <div class="content-mask-content">
            <div class="content-mask-icon">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-12 h-12">
                    <path fill-rule="evenodd" d="M5.166 2.621v.858c-1.035.148-2.059.33-3.071.543a.75.75 0 00-.584.859 6.937 6.937 0 006.222 6.222.75.75 0 00.859-.584 48.332 48.332 0 001.243-5.034 48.332 48.332 0 005.034-1.243.75.75 0 00.584-.859 6.937 6.937 0 00-6.222-6.222.75.75 0 00-.859.584 48.332 48.332 0 00-1.243 5.034zM13.5 2.25a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V3a.75.75 0 01.75-.75zM13.5 6.75a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V7.5a.75.75 0 01.75-.75zM13.5 11.25a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V12a.75.75 0 01.75-.75z" clip-rule="evenodd" />
                </svg>
            </div>
            <h3 class="text-xl font-semibold text-foreground mb-2">Konten Premium</h3>
            <p class="text-muted-foreground text-center mb-4">Untuk membaca {content_type} ini secara lengkap, Anda memerlukan langganan premium.</p>
            <div class="flex flex-col sm:flex-row gap-3 justify-center">
                <a href="/premium" class="button button-primary">
                    Berlangganan Premium
                </a>
                <a href="/login" class="button button-outline">
                    Login
                </a>
            </div>
        </div>
    </div>
    """ 

File: static/js/premium-content.js
/**
 * Premium Content Management
 * Handles premium content display, masking, and user interactions
 */

class PremiumContentManager {
    constructor() {
        this.isPremiumUser = window.currentUserId && window.currentUserRole;
        this.init();
    }

    init() {
        this.setupPremiumContentHandlers();
        this.setupContentMaskInteractions();
        this.setupPremiumStats();
    }

    setupPremiumContentHandlers() {
        // Handle premium content containers
        const premiumContainers = document.querySelectorAll('.premium-content-container');
        premiumContainers.forEach(container => {
            if (container.classList.contains('truncated')) {
                this.setupTruncatedContent(container);
            }
        });
    }

    setupTruncatedContent(container) {
        // Add fade effect to truncated content
        const content = container.querySelector('#article-body');
        if (content) {
            content.style.position = 'relative';
            content.style.overflow = 'hidden';
        }
    }

    setupContentMaskInteractions() {
        // Handle content mask interactions
        const contentMasks = document.querySelectorAll('.content-mask-overlay');
        contentMasks.forEach(mask => {
            this.setupMaskInteractions(mask);
        });
    }

    setupMaskInteractions(mask) {
        // Add click handlers for mask buttons
        const buttons = mask.querySelectorAll('.button');
        buttons.forEach(button => {
            button.addEventListener('click', (e) => {
                e.preventDefault();
                this.handlePremiumButtonClick(button);
            });
        });

        // Add hover effects
        mask.addEventListener('mouseenter', () => {
            mask.style.transform = 'scale(1.02)';
        });

        mask.addEventListener('mouseleave', () => {
            mask.style.transform = 'scale(1)';
        });
    }

    handlePremiumButtonClick(button) {
        const action = button.textContent.trim();
        
        if (action.includes('Berlangganan')) {
            // Redirect to premium page
            window.location.href = '/premium';
        } else if (action.includes('Login')) {
            // Redirect to login page
            window.location.href = '/login';
        }
    }

    setupPremiumStats() {
        // Animate premium stats
        const stats = document.querySelectorAll('.premium-content-stats .stat');
        stats.forEach((stat, index) => {
            setTimeout(() => {
                stat.style.opacity = '0';
                stat.style.transform = 'translateY(10px)';
                stat.style.transition = 'all 0.3s ease';
                
                setTimeout(() => {
                    stat.style.opacity = '1';
                    stat.style.transform = 'translateY(0)';
                }, 100);
            }, index * 100);
        });
    }

    // Method to reveal premium content (for testing or admin use)
    revealPremiumContent(contentId) {
        const container = document.querySelector(`#content-mask-${contentId}`);
        if (container) {
            container.style.display = 'none';
            
            // Remove truncated class from parent
            const contentContainer = container.closest('.premium-content-container');
            if (contentContainer) {
                contentContainer.classList.remove('truncated');
            }
        }
    }

    // Method to check if content is premium
    isContentPremium(contentId) {
        const container = document.querySelector(`#content-mask-${contentId}`);
        return container !== null;
    }

    // Method to get premium content statistics
    getPremiumContentStats() {
        const stats = {
            totalPremiumContent: document.querySelectorAll('.content-mask-overlay').length,
            totalTruncatedContent: document.querySelectorAll('.premium-content-container.truncated').length,
            userHasPremium: this.isPremiumUser
        };
        return stats;
    }
}

// Initialize premium content manager when DOM is loaded
document.addEventListener('DOMContentLoaded', function() {
    // Initialize premium content manager
    window.premiumContentManager = new PremiumContentManager();
    
    // Add global methods for external access
    window.revealPremiumContent = (contentId) => {
        if (window.premiumContentManager) {
            window.premiumContentManager.revealPremiumContent(contentId);
        }
    };
    
    window.isContentPremium = (contentId) => {
        if (window.premiumContentManager) {
            return window.premiumContentManager.isContentPremium(contentId);
        }
        return false;
    };
    
    window.getPremiumContentStats = () => {
        if (window.premiumContentManager) {
            return window.premiumContentManager.getPremiumContentStats();
        }
        return {};
    };
});

// Utility functions for premium content
const PremiumContentUtils = {
    // Check if user has premium access
    hasPremiumAccess: function() {
        return window.currentUserId && window.currentUserRole;
    },

    // Format premium content stats
    formatStats: function(stats) {
        return {
            totalWords: stats.total_words || 0,
            totalChars: stats.total_chars || 0,
            isPremium: stats.is_premium || false,
            userHasAccess: stats.user_has_access || false
        };
    },

    // Create premium notice element
    createPremiumNotice: function(contentType = 'artikel') {
        const notice = document.createElement('div');
        notice.className = 'premium-content-notice';
        notice.innerHTML = `
            <div class="flex items-center gap-3 mb-4">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-6 h-6 text-primary">
                    <path fill-rule="evenodd" d="M5.166 2.621v.858c-1.035.148-2.059.33-3.071.543a.75.75 0 00-.584.859 6.937 6.937 0 006.222 6.222.75.75 0 00.859-.584 48.332 48.332 0 001.243-5.034 48.332 48.332 0 005.034-1.243.75.75 0 00.584-.859 6.937 6.937 0 00-6.222-6.222.75.75 0 00-.859.584 48.332 48.332 0 00-1.243 5.034zM13.5 2.25a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V3a.75.75 0 01.75-.75zM13.5 6.75a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V7.5a.75.75 0 01.75-.75zM13.5 11.25a.75.75 0 01.75.75v.75a.75.75 0 01-1.5 0V12a.75.75 0 01.75-.75z" clip-rule="evenodd" />
                </svg>
                <h3 class="text-lg font-semibold text-foreground">Konten Premium</h3>
            </div>
            <p class="text-muted-foreground mb-4">Untuk membaca ${contentType} ini secara lengkap, Anda memerlukan langganan premium.</p>
            <div class="flex flex-wrap gap-3">
                <a href="/premium" class="button button-primary">
                    Berlangganan Premium
                </a>
                <a href="/login" class="button button-outline">
                    Login
                </a>
            </div>
        `;
        return notice;
    },

    // Show premium content notice
    showPremiumNotice: function(container, contentType = 'artikel') {
        const notice = this.createPremiumNotice(contentType);
        container.appendChild(notice);
    },

    // Hide premium content notice
    hidePremiumNotice: function(container) {
        const notice = container.querySelector('.premium-content-notice');
        if (notice) {
            notice.remove();
        }
    }
};

// Export for module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { PremiumContentManager, PremiumContentUtils };
} 